File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/tasks/geodatabase/ConvertGeodatabase.java
package com.valentingrigorean.arcgis_maps_flutter.tasks.geodatabase;

import com.esri.arcgisruntime.data.FeatureEditResult;
import com.esri.arcgisruntime.data.SyncModel;
import com.esri.arcgisruntime.tasks.geodatabase.GenerateGeodatabaseParameters;
import com.esri.arcgisruntime.tasks.geodatabase.GenerateLayerOption;
import com.esri.arcgisruntime.tasks.geodatabase.GeodatabaseDeltaInfo;
import com.esri.arcgisruntime.tasks.geodatabase.SyncGeodatabaseParameters;
import com.esri.arcgisruntime.tasks.geodatabase.SyncLayerOption;
import com.esri.arcgisruntime.tasks.geodatabase.SyncLayerResult;
import com.esri.arcgisruntime.tasks.geodatabase.UtilityNetworkSyncMode;
import com.valentingrigorean.arcgis_maps_flutter.Convert;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ConvertGeodatabase extends Convert {

    public static Object geodatabaseDeltaInfoToJson(GeodatabaseDeltaInfo geodatabaseDeltaInfo) {
        final HashMap<String, Object> data = new HashMap<>(4);
        if (geodatabaseDeltaInfo.getDownloadDeltaPath() != null) {
            data.put("downloadDeltaFileUrl", geodatabaseDeltaInfo.getDownloadDeltaPath());
        }
        data.put("featureServiceUrl", geodatabaseDeltaInfo.getFeatureServiceUrl());
        data.put("geodatabaseFileUrl", geodatabaseDeltaInfo.getGeodatabasePath());
        if (geodatabaseDeltaInfo.getUploadDeltaPath() != null) {
            data.put("uploadDeltaFileUrl", geodatabaseDeltaInfo.getUploadDeltaPath());
        }
        return data;
    }


    public static Object syncGeodatabaseParametersToJson(SyncGeodatabaseParameters syncGeodatabaseParameters) {
        final HashMap<String, Object> data = new HashMap<>(4);
        data.put("keepGeodatabaseDeltas", syncGeodatabaseParameters.isKeepGeodatabaseDeltas());
        data.put("geodatabaseSyncDirection", syncDirectionToJson(syncGeodatabaseParameters.getSyncDirection()));
        data.put("layerOptions", syncLayerOptionsToJson(syncGeodatabaseParameters.getLayerOptions()));
        data.put("rollbackOnFailure", syncGeodatabaseParameters.isRollbackOnFailure());
        return data;
    }

    public static SyncGeodatabaseParameters toSyncGeodatabaseParameters(Object o) {
        final Map<?, ?> data = toMap(o);
        final SyncGeodatabaseParameters syncGeodatabaseParameters = new SyncGeodatabaseParameters();
        syncGeodatabaseParameters.setKeepGeodatabaseDeltas(toBoolean(data.get("keepGeodatabaseDeltas")));
        syncGeodatabaseParameters.setSyncDirection(toSyncDirection(data.get("geodatabaseSyncDirection")));
        syncGeodatabaseParameters.getLayerOptions().addAll(toSyncLayerOptions(data.get("layerOptions")));
        syncGeodatabaseParameters.setRollbackOnFailure(toBoolean(data.get("rollbackOnFailure")));
        return syncGeodatabaseParameters;
    }

    private static Collection<SyncLayerOption> toSyncLayerOptions(Object layerOptions) {
        final List<?> data = toList(layerOptions);
        final ArrayList<SyncLayerOption> syncLayerOptions = new ArrayList<>(data.size());
        for (final Object layerOption : data) {
            syncLayerOptions.add(toSyncLayerOption(layerOption));
        }
        return syncLayerOptions;
    }

    public static Object syncLayerResultsToJson(List<SyncLayerResult> syncResults) {
        final ArrayList<Object> data = new ArrayList<>(syncResults.size());
        for (final SyncLayerResult layerResult : syncResults) {
            data.add(syncLayerResultToJson(layerResult));
        }
        return data;
    }

    public static Object generateGeodatabaseParametersToJson(GenerateGeodatabaseParameters parameters) {
        final HashMap<String, Object> json = new HashMap<>();
        json.put("attachmentSyncDirection", attachmentSyncDirectionToJson(parameters.getAttachmentSyncDirection()));
        if (parameters.getExtent() != null) {
            json.put("extent", geometryToJson(parameters.getExtent()));
        }
        final ArrayList<Object> layersOptions = new ArrayList<>();
        for (GenerateLayerOption option : parameters.getLayerOptions()) {
            layersOptions.add(generateLayerOptionToJson(option));
        }
        json.put("layerOptions", layersOptions);
        if (parameters.getOutSpatialReference() != null) {
            json.put("outSpatialReference", spatialReferenceToJson(parameters.getOutSpatialReference()));
        }
        json.put("returnAttachments", parameters.isReturnAttachments());
        json.put("shouldSyncContingentValues", parameters.getSyncContingentValues());
        json.put("syncModel", parameters.getSyncModel().ordinal());
        json.put("utilityNetworkSyncMode", parameters.getUtilityNetworkSyncMode().ordinal());
        return json;
    }

    public static GenerateGeodatabaseParameters toGenerateGeodatabaseParameters(Object json) {
        final Map<?, ?> data = toMap(json);
        final Object attachmentSyncDirection = data.get("attachmentSyncDirection");
        final Object extent = data.get("extent");
        final List<?> layerOptions = toList(data.get("layerOptions"));
        final Object outSpatialReference = data.get("outSpatialReference");
        final Object returnAttachments = data.get("returnAttachments");
        final Object shouldSyncContingentValues = data.get("shouldSyncContingentValues");
        final Object syncModel = data.get("syncModel");
        final Object utilityNetworkSyncMode = data.get("utilityNetworkSyncMode");

        final GenerateGeodatabaseParameters parameters = new GenerateGeodatabaseParameters();

        parameters.setAttachmentSyncDirection(toAttachmentSyncDirection(attachmentSyncDirection));
        if (extent != null) {
            parameters.setExtent(toGeometry(extent));
        }
        for (final Object layerOption : layerOptions) {
            parameters.getLayerOptions().add(toGenerateLayerOption(layerOption));
        }
        if (outSpatialReference != null) {
            parameters.setOutSpatialReference(toSpatialReference(outSpatialReference));
        }
        parameters.setReturnAttachments(toBoolean(returnAttachments));
        parameters.setSyncContingentValues(toBoolean(shouldSyncContingentValues));
        parameters.setSyncModel(SyncModel.values()[toInt(syncModel)]);
        parameters.setUtilityNetworkSyncMode(UtilityNetworkSyncMode.values()[toInt(utilityNetworkSyncMode)]);
        return parameters;
    }

    private static Object syncLayerOptionsToJson(List<SyncLayerOption> syncLayerOptions) {
        final ArrayList<Object> data = new ArrayList<>(syncLayerOptions.size());
        for (final SyncLayerOption layerOption : syncLayerOptions) {
            data.add(syncLayerOptionToJson(layerOption));
        }
        return data;
    }

    private static Object syncLayerOptionToJson(SyncLayerOption syncLayerOption) {
        final HashMap<String, Object> data = new HashMap<>(2);
        data.put("layerId", syncLayerOption.getLayerId());
        data.put("syncDirection", syncDirectionToJson(syncLayerOption.getSyncDirection()));
        return data;
    }

    private static SyncLayerOption toSyncLayerOption(Object json) {
        final Map<?, ?> data = toMap(json);
        final Object layerId = data.get("layerId");
        final Object syncDirection = data.get("syncDirection");
        return new SyncLayerOption(toInt(layerId), toSyncDirection(syncDirection));
    }

    private static Object syncLayerResultToJson(SyncLayerResult syncResult) {
        final HashMap<String, Object> json = new HashMap<>(3);

        final ArrayList<Object> editsResults = new ArrayList<>(syncResult.getEditResults().size());
        for (final FeatureEditResult editResult : syncResult.getEditResults()) {
            editsResults.add(featureEditResultToJson(editResult));
        }
        json.put("editResults", editsResults);
        json.put("layerId", syncResult.getLayerId());
        json.put("tableName", syncResult.getTableName());
        return json;
    }


    private static Object generateLayerOptionToJson(GenerateLayerOption layerOption) {
        final HashMap<String, Object> json = new HashMap<>(5);
        json.put("layerId", layerOption.getLayerId());
        json.put("includeRelated", layerOption.isIncludeRelated());
        json.put("queryOption", layerOption.getQueryOption().ordinal());
        json.put("useGeometry", layerOption.isUseGeometry());
        json.put("whereClause", layerOption.getWhereClause());
        return json;
    }

    private static GenerateLayerOption toGenerateLayerOption(Object json) {
        final Map<?, ?> data = toMap(json);
        final GenerateLayerOption layerOption = new GenerateLayerOption();
        layerOption.setLayerId(Convert.toLong(data.get("layerId")));
        layerOption.setIncludeRelated(toBoolean(data.get("includeRelated")));
        layerOption.setQueryOption(GenerateLayerOption.QueryOption.values()[Convert.toInt(data.get("queryOption"))]);
        layerOption.setUseGeometry(toBoolean(data.get("useGeometry")));
        layerOption.setWhereClause((String) data.get("whereClause"));
        return layerOption;
    }

    private static int attachmentSyncDirectionToJson(GenerateGeodatabaseParameters.AttachmentSyncDirection direction) {
        switch (direction) {
            case UPLOAD:
                return 1;
            case BIDIRECTIONAL:
                return 2;
            default:
                return 0;
        }
    }

    private static GenerateGeodatabaseParameters.AttachmentSyncDirection toAttachmentSyncDirection(Object direction) {
        switch (toInt(direction)) {
            case 1:
                return GenerateGeodatabaseParameters.AttachmentSyncDirection.UPLOAD;
            case 2:
                return GenerateGeodatabaseParameters.AttachmentSyncDirection.BIDIRECTIONAL;
            default:
                return GenerateGeodatabaseParameters.AttachmentSyncDirection.NONE;
        }
    }

}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/tasks/geodatabase/ConvertGeodatabase.kt
package com.valentingrigorean.arcgis_maps_flutter.tasks.geodatabase

import com.esri.arcgisruntime.data.SyncModel
import com.esri.arcgisruntime.tasks.geodatabase.GenerateGeodatabaseParameters
import com.esri.arcgisruntime.tasks.geodatabase.GenerateLayerOption
import com.esri.arcgisruntime.tasks.geodatabase.GeodatabaseDeltaInfo
import com.esri.arcgisruntime.tasks.geodatabase.SyncGeodatabaseParameters
import com.esri.arcgisruntime.tasks.geodatabase.SyncLayerOption
import com.esri.arcgisruntime.tasks.geodatabase.SyncLayerResult
import com.esri.arcgisruntime.tasks.geodatabase.UtilityNetworkSyncMode
import com.valentingrigorean.arcgis_maps_flutter.Convert
import com.valentingrigorean.arcgis_maps_flutter.utils.toMap

object ConvertGeodatabase : Convert() {
    fun geodatabaseDeltaInfoToJson(geodatabaseDeltaInfo: GeodatabaseDeltaInfo): Any {
        val data = HashMap<String, Any>(4)
        if (geodatabaseDeltaInfo.downloadDeltaPath != null) {
            data["downloadDeltaFileUrl"] = geodatabaseDeltaInfo.downloadDeltaPath
        }
        data["featureServiceUrl"] = geodatabaseDeltaInfo.featureServiceUrl
        data["geodatabaseFileUrl"] = geodatabaseDeltaInfo.geodatabasePath
        if (geodatabaseDeltaInfo.uploadDeltaPath != null) {
            data["uploadDeltaFileUrl"] = geodatabaseDeltaInfo.uploadDeltaPath
        }
        return data
    }

    fun syncGeodatabaseParametersToJson(syncGeodatabaseParameters: SyncGeodatabaseParameters): Any {
        val data = HashMap<String, Any>(4)
        data["keepGeodatabaseDeltas"] = syncGeodatabaseParameters.isKeepGeodatabaseDeltas
        data["geodatabaseSyncDirection"] =
            Convert.Companion.syncDirectionToJson(
                syncGeodatabaseParameters.syncDirection
            )
        data["layerOptions"] =
            syncLayerOptionsToJson(syncGeodatabaseParameters.layerOptions)
        data["rollbackOnFailure"] = syncGeodatabaseParameters.isRollbackOnFailure
        return data
    }

    fun toSyncGeodatabaseParameters(o: Any?): SyncGeodatabaseParameters {
        val data: Map<*, *> = Convert.Companion.toMap(
            o!!
        )
        val syncGeodatabaseParameters = SyncGeodatabaseParameters()
        syncGeodatabaseParameters.isKeepGeodatabaseDeltas = Convert.Companion.toBoolean(
            data["keepGeodatabaseDeltas"]!!
        )
        syncGeodatabaseParameters.syncDirection = Convert.Companion.toSyncDirection(
            data["geodatabaseSyncDirection"]
        )
        syncGeodatabaseParameters.layerOptions.addAll(toSyncLayerOptions(data["layerOptions"]))
        syncGeodatabaseParameters.isRollbackOnFailure = Convert.Companion.toBoolean(
            data["rollbackOnFailure"]!!
        )
        return syncGeodatabaseParameters
    }

    private fun toSyncLayerOptions(layerOptions: Any?): Collection<SyncLayerOption> {
        val data: List<*> = Convert.Companion.toList(
            layerOptions!!
        )
        val syncLayerOptions = ArrayList<SyncLayerOption>(data.size)
        for (layerOption in data) {
            syncLayerOptions.add(toSyncLayerOption(layerOption))
        }
        return syncLayerOptions
    }

    fun syncLayerResultsToJson(syncResults: List<SyncLayerResult>): Any {
        val data = ArrayList<Any>(syncResults.size)
        for (layerResult in syncResults) {
            data.add(syncLayerResultToJson(layerResult))
        }
        return data
    }

    fun generateGeodatabaseParametersToJson(parameters: GenerateGeodatabaseParameters): Any {
        val json = HashMap<String, Any?>()
        json["attachmentSyncDirection"] =
            attachmentSyncDirectionToJson(parameters.attachmentSyncDirection)
        if (parameters.extent != null) {
            json["extent"] = Convert.Companion.geometryToJson(parameters.extent)
        }
        val layersOptions = ArrayList<Any>()
        for (option in parameters.layerOptions) {
            layersOptions.add(generateLayerOptionToJson(option))
        }
        json["layerOptions"] = layersOptions
        if (parameters.outSpatialReference != null) {
            json["outSpatialReference"] =
                Convert.Companion.spatialReferenceToJson(parameters.outSpatialReference)
        }
        json["returnAttachments"] = parameters.isReturnAttachments
        json["shouldSyncContingentValues"] = parameters.syncContingentValues
        json["syncModel"] = parameters.syncModel.ordinal
        json["utilityNetworkSyncMode"] = parameters.utilityNetworkSyncMode.ordinal
        return json
    }

    fun toGenerateGeodatabaseParameters(json: Any?): GenerateGeodatabaseParameters {
        val data: Map<*, *> = Convert.Companion.toMap(
            json!!
        )
        val attachmentSyncDirection = data["attachmentSyncDirection"]
        val extent = data["extent"]
        val layerOptions: List<*> = Convert.Companion.toList(
            data["layerOptions"]!!
        )
        val outSpatialReference = data["outSpatialReference"]
        val returnAttachments = data["returnAttachments"]
        val shouldSyncContingentValues = data["shouldSyncContingentValues"]
        val syncModel = data["syncModel"]
        val utilityNetworkSyncMode = data["utilityNetworkSyncMode"]
        val parameters = GenerateGeodatabaseParameters()
        parameters.attachmentSyncDirection = toAttachmentSyncDirection(attachmentSyncDirection)
        if (extent != null) {
            parameters.extent = Convert.Companion.toGeometry(extent)
        }
        for (layerOption in layerOptions) {
            parameters.layerOptions.add(toGenerateLayerOption(layerOption))
        }
        if (outSpatialReference != null) {
            parameters.outSpatialReference =
                Convert.Companion.toSpatialReference(outSpatialReference)
        }
        parameters.isReturnAttachments = Convert.Companion.toBoolean(
            returnAttachments!!
        )
        parameters.syncContingentValues = Convert.Companion.toBoolean(
            shouldSyncContingentValues!!
        )
        parameters.syncModel = SyncModel.values()[Convert.Companion.toInt(syncModel)]
        parameters.utilityNetworkSyncMode =
            UtilityNetworkSyncMode.values()[Convert.Companion.toInt(utilityNetworkSyncMode)]
        return parameters
    }

    private fun syncLayerOptionsToJson(syncLayerOptions: List<SyncLayerOption>): Any {
        val data = ArrayList<Any>(syncLayerOptions.size)
        for (layerOption in syncLayerOptions) {
            data.add(syncLayerOptionToJson(layerOption))
        }
        return data
    }

    private fun syncLayerOptionToJson(syncLayerOption: SyncLayerOption): Any {
        val data = HashMap<String, Any>(2)
        data["layerId"] = syncLayerOption.layerId
        data["syncDirection"] =
            Convert.Companion.syncDirectionToJson(syncLayerOption.syncDirection)
        return data
    }

    private fun toSyncLayerOption(json: Any): SyncLayerOption {
        val data: Map<*, *> = Convert.Companion.toMap(json)
        val layerId = data["layerId"]
        val syncDirection = data["syncDirection"]
        return SyncLayerOption(
            Convert.Companion.toInt(layerId).toLong(),
            Convert.Companion.toSyncDirection(syncDirection)
        )
    }

    private fun syncLayerResultToJson(syncResult: SyncLayerResult): Any {
        val json = HashMap<String, Any>(3)
        val editsResults = ArrayList<Any>(syncResult.editResults.size)
        for (editResult in syncResult.editResults) {
            editsResults.add(Convert.Companion.featureEditResultToJson(editResult))
        }
        json["editResults"] = editsResults
        json["layerId"] = syncResult.layerId
        json["tableName"] = syncResult.tableName
        return json
    }

    private fun generateLayerOptionToJson(layerOption: GenerateLayerOption): Any {
        val json = HashMap<String, Any>(5)
        json["layerId"] = layerOption.layerId
        json["includeRelated"] = layerOption.isIncludeRelated
        json["queryOption"] = layerOption.queryOption.ordinal
        json["useGeometry"] = layerOption.isUseGeometry
        json["whereClause"] = layerOption.whereClause
        return json
    }

    private fun toGenerateLayerOption(json: Any): GenerateLayerOption {
        val data: Map<*, *> = Convert.Companion.toMap(json)
        val layerOption = GenerateLayerOption()
        layerOption.layerId = Convert.Companion.toLong(
            data["layerId"]
        )
        layerOption.isIncludeRelated = Convert.Companion.toBoolean(
            data["includeRelated"]!!
        )
        layerOption.queryOption = GenerateLayerOption.QueryOption.values()[Convert.Companion.toInt(
            data["queryOption"]
        )]
        layerOption.isUseGeometry = Convert.Companion.toBoolean(
            data["useGeometry"]!!
        )
        layerOption.whereClause = data["whereClause"] as String?
        return layerOption
    }

    private fun attachmentSyncDirectionToJson(direction: GenerateGeodatabaseParameters.AttachmentSyncDirection): Int {
        return when (direction) {
            GenerateGeodatabaseParameters.AttachmentSyncDirection.UPLOAD -> 1
            GenerateGeodatabaseParameters.AttachmentSyncDirection.BIDIRECTIONAL -> 2
            else -> 0
        }
    }

    private fun toAttachmentSyncDirection(direction: Any?): GenerateGeodatabaseParameters.AttachmentSyncDirection {
        return when (Convert.Companion.toInt(direction)) {
            1 -> GenerateGeodatabaseParameters.AttachmentSyncDirection.UPLOAD
            2 -> GenerateGeodatabaseParameters.AttachmentSyncDirection.BIDIRECTIONAL
            else -> GenerateGeodatabaseParameters.AttachmentSyncDirection.NONE
        }
    }
}